app/scripts/metamask-controller.js

```diff
@@ -4386,117 +4379,52 @@ export default class MetamaskController extends EventEmitter {
   // Log blocks
 
   /**
-   * A method for selecting a custom URL for an ethereum RPC provider and updating it
+   * Adds a network configuration if the rpcUrl is not already present on an
+   * existing network configuration. Otherwise updates the entry with the matching rpcUrl.
    *
-   * @param {string} rpcUrl - A URL for a valid Ethereum RPC API.
-   * @param {string} chainId - The chainId of the selected network.
-   * @param {string} ticker - The ticker symbol of the selected network.
-   * @param {string} [nickname] - Nickname of the selected network.
-   * @param {object} [rpcPrefs] - RPC preferences.
-   * @param {string} [rpcPrefs.blockExplorerUrl] - URL of block explorer for the chain.
-   * @returns {Promise<string>} The RPC Target URL confirmed.
+   * @param config - The network configuration.
+   * @param config.rpcUrl - The network configuration RPC URL.
+   * @param config.chainId - The chain ID of the network, as per EIP-155.
+   * @param config.ticker - Currency ticker.
+   * @param config.chainName - Personalized network name.
+   * @param config.rpcPrefs - Personalized preferences.
+   * @returns id for the added or updated network configuration
    */
-  async updateAndSetCustomRpc(
+  async upsertAndSetActiveNetwork({
     rpcUrl,
     chainId,
     ticker = 'ETH',
-    nickname,
+    chainName,
     rpcPrefs,
-  ) {
-    this.networkController.setRpcTarget(
+  }) {
+    const networkConfigurationId = await this.upsertNetworkConfiguration({
       rpcUrl,
       chainId,
       ticker,
-      nickname,
-      rpcPrefs,
-    );
-    await this.preferencesController.upsertToFrequentRpcList(
-      rpcUrl,
-      chainId,
-      ticker,
-      nickname,
+      chainName,
       rpcPrefs,
```

Comment:
The `upsertNetworkConfiguration` method accepts a `blockExplorerUrl` property here instead of `rpcPrefs`

Reply:
I see that the same problem exists in `actions.js` - it's calling `upsertNetworkConfiguration` with the params expected by the network controller version of that method, but really it's the metamask-controller version being called.

Maybe they can be updated to accept the same parameters to avoid this confusion? Or one of them renamed.

Reply:
Alternatively we could remove the metamask-controller version altogether, and bake in the event tracking somehow. I see that elsewhere we're passing in a `trackMetaMetricsEvent` function to controller constructors, we could do that here too.

Reply:
Good catch! I feel like we ought to pass in `trackMetaMetricsEvent` and try to declutter the `metamask-controller` with these relayed calls that just tack on event tracking where possible!

Reply:
ah the problem here is that the `MetaMetricsController` relies on the `NetworkController` so we can't instantiate it until after. So we can't pass the `trackMetaMetricsEvent` function into the constructor.

Reply:
Fixed here: https://github.com/MetaMask/metamask-extension/pull/17421/commits/dccca031bdf9a943ef61b46c2c97b40ae6a28245

Reply:
We could pass it into the constructor by passing it in wrapped in a function.

e.g. this won't work, for the reason you described:

```
      trackMetaMetricsEvent: this.metaMetricsController.trackEvent.bind(
        this.metaMetricsController,
      ),
```

But this will:

```
      trackMetaMetricsEvent: (...args) => this.metaMetricsController.trackEvent(...args),
```

Reply:
Fixing it the way you have in that commit is fine though :+1: